File: auto-reapair.py
from requests import get,post;import os;from time import sleep;import json
comando = os.system

#Instalando Ou Verificando ADB-TOOLS

def LookAndInstall():
    try:
        fastboot = os.popen('fastboot -help').readlines(); fastboot[0]
        print('[V] ADB-Tools already installed....')
    except:
        try:
            print('[V] Installing ADB-Tools and our dependencies...')
            sleep(1.5)
            comando('sudo apt install android-tools-adb;sudo apt install android-tools-fastboot')
            fastboot = os.popen('fastboot -help').readlines(); fastboot[0]
            comando('clear')
            print("[ADB-Tools] Successfully installed!")   
        except:
            print("[X] An error occurred installing ADB-Tools, try again")
            exit()


#Achando a ROM para o dispositivo

def FindRom(dispositivo):
    dispositivo = dispositivo.upper()
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.114 Safari/537.36',
        'Content-Type': 'application/json;charset=UTF-8'
    }
    data = '{"action":"get","search":{"href":"/firmware/","pattern":"'+ dispositivo +'","ignorecase":true}}'
    req = post('https://mirrors.lolinet.com/firmware/?', headers=headers, data=data)
    a = json.loads(req.text)
    txt_brt = a['search']
    for versao in txt_brt:
        ultima_versao = versao
    arquivo_para_baixar = ultima_versao['href']
    arquivo = arquivo_para_baixar.split(f"{dispositivo}",1)[1];arquivo = dispositivo + arquivo
    try:
        comando(f'sudo wget https://mirrors.lolinet.com{arquivo_para_baixar}')
        diretorio_arquivo = arquivo.replace('.zip', '')
        comando(f'sudo unzip {arquivo}; cd {diretorio_arquivo}')
        print('[V] ROM successfully downloaded and uncompressed!')
        sleep(0.5)
    except:
        print('[X] An error occurred while downloading/decompressing the ROM')
        exit()
    

#Detectando e instalando 

def Detect_And_Install():
    try:
        fastboot = os.popen('fastboot devices').readlines()
    except:
        print('[X] There was an error in the installation process, try again')
    if 'fastboot' in fastboot:
        a = 0
        os.system('clear')
        print("[V] Aparelho Detectado!")
        print('[V] Inserindo a rom....')
        sleep(2)
        comando('fastboot oem fb_mode_set')
        sleep(0.3)
        comando('fastboot flash partition gpt.bin')
        comando('fastboot flash bootloader bootloader.img')
        comando('fastboot flash logo logo.bin')
        comando('fastboot flash boot boot.img')
        comando('fastboot flash recovery recovery.img')
        comando('fastboot flash dsp adspso.bin')
        comando('fastboot flash oem oem.img')
        while a <= 11:
            comando(f'fastboot flash system system.img_sparsechunk.{a}')
            a += 1
        comando('fastboot flash modem NON-HLOS.bin')
        comando('fastboot erase modemst1')
        comando('fastboot erase modemst2')
        comando('fastboot flash fsg fsg.mbn')
        comando('fastboot erase cache ')
        comando('fastboot erase userdata')
        comando('fastboot erase customize')
        comando('fastboot erase clogo')
        comando('fastboot oem fb_mode_clear')
        comando('fastboot flash logo logo.bin')
        comando('fastboot reboot')
        comando('clear')
        print('[V] Restarting device....')
        sleep(3)
        print('[V] ROM successfully installed!')
    else:
        print("[X] Please connect your device in FastBoot mode.")
        exit()


#Func Main

def Main():
    LookAndInstall()
    q = input('[|] Do you want to download an appropriate ROM for your device? [Y/N]: ')
    if 'Y' in q or 'y' in q:
        try:
            dispositivo = input('[V] Please, enter your device: ')
            FindRom(str(dispositivo))
            Detect_And_Install()
        except:
            print('Error.')
    elif 'N' in q or 'n' in q:
        Detect_And_Install()
    else:
        print('[X] Just answer with Y or N')
Main()

File: auto-reapair.c
#include <stdio.h>  
#include <stdlib.h>
#include <string.h> 
#include <unistd.h> 

// Verifica o comando fastboot 
int Retorno_Verf(void){
  
    FILE *cmd = popen("fastboot -help", "r");
    int retorno;
    char str_detect[6] = "--help";
    char result[500]={0x0};

    while (fgets(result, sizeof(result), cmd) !=NULL)
    system("clear");
    retorno = strncmp(result, str_detect, 6);
    return retorno;
}

// Instalando Ou Verificando ADB-TOOLS
int LookAndInstall(void){

    if (Retorno_Verf() > -15){
        printf("[V] ADB-Tools already installed....\n");
        return 0;
    }
    else
        printf("[V] Installing ADB-Tools and our dependencies...\n");
        sleep(3);
        system("sudo apt install android-tools-adb;sudo apt install android-tools-fastboot");
        system("clear");
        printf("[V] Verifying installation.....\n");
        if (Retorno_Verf() > -15){
            printf("[ADB-Tools] Successfully installed!\n");
            sleep(1.5);
            return 0;
        }
        else{
            printf("[X] An Error Ocurred.\n");
            exit(0);
        }
}


// Olhando para o comando de dispositivos
int Devices_Look(){

    FILE *cmd = popen("fastboot devices", "r");
    char str_detect[6] = "0000";
    int retorno;
    char result[500]={0x0};
    while (fgets(result, sizeof(result), cmd) !=NULL)
    sleep(0.5);
    retorno = strncmp(result, str_detect, 6);
    return retorno;
}

// Detectando dispositivos conectados
int Detect(){
    if (Devices_Look() > 0){
        printf("[V] Detected Device!\n");     
       return 1;
    } else {
        printf("[X] Please, connect your device in fastboot mode.\n");
        while(Devices_Look() < 0){
            if (Devices_Look() > 0){
                break;
            }
        }
        printf("[V] Detected Device!\n");
    }
}

// Instalando a rom
// (Certifique-se de que está sendo executado na mesma pasta em que a rom descompactada se localiza)
int Install(){
    Detect();
    printf("[V] Installing ROM.....\n");
    printf("[|] Make Sure That the Script is running on same paste of the ROM.....");
    sleep(4.5);
    system("fastboot oem fb_ mode_set");
    sleep(2); // Previnir Erros.
    system("fastboot flash partition gpt.bin");
    system("fastboot flash bootloader bootloader.img");
    system("fastboot flash logo logo.bin");
    system("fastboot flash boot boot.img");
    system("fastboot flash recovery recovery.img");
    system("fastboot flash dsp adspso.bin");
    system("fastboot flash oem oem.img");
    int a = 0;
    while (a <= 11){
        char comando[70];
        char string[50] = "fastboot flash system system.img_sparsechunk.";
        sprintf(comando,"%s%d", string, a);
        system(comando);
        a += 1; 
    }
    sleep(1); // Previnir Erros.
    system("fastboot flash modem NON-HLOS.bin");
    system("fastboot erase modemst1");
    system("fastboot erase modemst2");
    system("fastboot flash fsg fsg.mbn");
    system("fastboot erase cache");
    system("fastboot erase userdata");
    system("fastboot erase customize");
    system("fastboot erase clogo");
    system("fastboot oem fb_mode_clear");
    system("fastboot flash logo logo.bin");
    system("fastboot reboot");
    system("clear");
    printf("[V] Restarting device....\n");
    sleep(4); // Previnir do script atrapalhar o reboot
    printf("[V] ROM successfully installed!\n");
}

// Func Main, inicia todo o processo.
int main(void){
    LookAndInstall();
    Install();
    return 0;
}


// By BlackReaper.
